File: code/jvm/repository/src/main/java/pt/isel/repositories/channels/ChannelMemberRepository.java
package pt.isel.repositories.channels;

import pt.isel.domain.channels.AccessType;
import pt.isel.domain.channels.Channel;
import pt.isel.domain.channels.ChannelMember;
import pt.isel.domain.users.UserInfo;
import pt.isel.repositories.Repository;

import java.util.List;

public interface ChannelMemberRepository extends Repository<ChannelMember> {
    ChannelMember addUserToChannel(UserInfo userInfo, Channel channel, AccessType accessType);

    ChannelMember findUserInChannel(Long userId, Long channelId);

    List<ChannelMember> findAllChannelsForUser(Long userId, int limit, int offset);

    void removeUserFromChannel(Long userId, Long channelId);

    List<ChannelMember> findAllMembersInChannel(Long channelId, int limit, int offset);
}

File: code/jvm/repository/src/main/kotlin/pt/isel/repositories/channels/ChannelMemberRepository.kt
package pt.isel.repositories.channels

import pt.isel.domain.channels.AccessType
import pt.isel.domain.channels.Channel
import pt.isel.domain.channels.ChannelMember
import pt.isel.domain.users.UserInfo
import pt.isel.repositories.Repository

/** Repository interface for managing channel members, extends the generic Repository */
interface ChannelMemberRepository : Repository<ChannelMember> {
    fun addUserToChannel(
        userInfo: UserInfo,
        channel: Channel,
        accessType: AccessType,
    ): ChannelMember

    fun findUserInChannel(
        userId: Long,
        channelId: Long,
    ): ChannelMember?

    fun findAllChannelsForUser(
        userId: Long,
        limit: Int = 50,
        offset: Int = 0,
    ): List<ChannelMember>

    fun removeUserFromChannel(
        userId: Long,
        channelId: Long,
    )

    fun findAllMembersInChannel(
        channelId: Long,
        limit: Int = 50,
        offset: Int = 0,
    ): List<ChannelMember>

    fun removeAllMembershipsForUser(userId: Long)
}
